server/src/main/java/org/elasticsearch/snapshots/RestoreService.java

```diff
@@ -1613,7 +1613,17 @@ private IndexMetadata convertLegacyIndex(IndexMetadata snapshotIndexMetadata) {
         Map<String, Object> newMapping = new LinkedHashMap<>();
         newMapping.put(mappingMetadata.type(), newMappingSource);
         // TODO: _routing? Perhaps we don't need to obey any routing here as stuff is read-only anyway and get API will be disabled
-        return IndexMetadata.builder(snapshotIndexMetadata).putMapping(new MappingMetadata(mappingMetadata.type(), newMapping)).build();
+        return IndexMetadata.builder(snapshotIndexMetadata)
+            .putMapping(new MappingMetadata(mappingMetadata.type(), newMapping))
+            .settings(
+                Settings.builder()
+                    .put(snapshotIndexMetadata.getSettings())
+                    .put(
+                        IndexMetadata.SETTING_INDEX_VERSION_COMPATIBILITY.getKey(),
+                        clusterState.getNodes().getSmallestNonClientNodeVersion()
```

Comment:
Can we use `getMinNodeVersion` or fail if `getSmallestNonClientNodeVersion != getMinNodeVersion`? The version barrier only has guarantees on `getMinNodeVersion` and it would be nice to rely on it here.

Reply:
I've used the same logic as we use for index creation, as this felt similar in nature. I agree that the interplay with the version barrier is confusing, but it's something to be addressed everywhere then (index creation is probably more frequent than what's being done here).

Reply:
I prefer to leave this to a follow-up, as it's a more general change to make (to index creation as well).